File: lib/algo/DoubleVector.lua
--- `algo.DoubleVector` class.
-- A vector with double-precision floating points as it elements, which is
-- implemented in C.  Implicitly called when the elements are numbers.
-- [ffi from LuaJIT](http://luajit.org/ext_ffi.html) is used for binding.
-- @classmod DoubleVector
local ffi = require "ffi"
local util = require "algo.Util"

ffi.cdef([[
  typedef struct {
  size_t size;
  double* vec;
  } DoubleVector;

  DoubleVector* double_vector_new(size_t);
  size_t double_vector_size(DoubleVector*);
  double double_vector_get(DoubleVector*, size_t);
  void double_vector_set(DoubleVector*, size_t, double);
  double double_vector_magnitude(DoubleVector*);
  int double_vector_equal(DoubleVector*, DoubleVector*);
  DoubleVector* double_vector_add(DoubleVector*, DoubleVector*);
  DoubleVector* double_vector_scalar_add(DoubleVector*, double);
  DoubleVector* double_vector_sub(DoubleVector*, DoubleVector*);
  DoubleVector* double_vector_scalar_sub_first(double, DoubleVector*);
  DoubleVector* double_vector_scalar_sub_second(DoubleVector*, double);
  DoubleVector* double_vector_mul(DoubleVector*, DoubleVector*);
  DoubleVector* double_vector_scalar_mul(DoubleVector*, double);
  DoubleVector* double_vector_div(DoubleVector*, DoubleVector*);
  DoubleVector* double_vector_scalar_div_first(double, DoubleVector*);
  DoubleVector* double_vector_scalar_div_second(DoubleVector*, double);
  DoubleVector* double_vector_pow(DoubleVector*, DoubleVector*);
  DoubleVector* double_vector_scalar_pow_first(double, DoubleVector*);
  DoubleVector* double_vector_scalar_pow_second(DoubleVector*, double);
  double double_vector_dot(DoubleVector*, DoubleVector*);
  void double_vector_error(DoubleVector*, const char*);
  void double_vector_free(DoubleVector*);
  char* utoa(unsigned);
  ]])

local cvector = nil
local os_type, _ = util.get_os()

-- Refactor it later.
local version = _VERSION:match("%d+%.%d+")
local prefix = os.getenv("PREFIX") or os.getenv("HOME")
local rockpath_unix = nil
local rockpath_win = nil
if prefix == os.getenv("HOME") then
  rockpath_unix = "/.luarocks/lib/lua/" .. version .. "/"
  rockpath_win = "\\.luarocks\\lib\\lua\\" .. version .. "\\"
else
  rockpath_unix = "/lib/luarocks/lib/lua/" .. version .. "/"
  rockpath_win = "\\lib\\luarocks\\lib\\lua\\" .. version .. "\\"
end

if os_type == "Windows" then
  cvector = ffi.load(prefix .. rockpath_win .. "libdoubleVector.dll")
elseif os_type == "Mac" then
  cvector = ffi.load(prefix .. rockpath_unix .. "libdoubleVector.dylib")
elseif os_type == "Linux" then
  cvector = ffi.load(prefix .. rockpath_unix .. "libdoubleVector.so")
else
  error("Unsupported platform")
end

-- local cvector = ffi.load("libdoubleVector")

local DoubleVector = {}
package.loaded['DoubleVector'] = DoubleVector

DoubleVector.__index = DoubleVector

--- Automatically called when the object is recycled.
DoubleVector.__gc = function (o)
  cvector.double_vector_free(o)
end

--- Check whether two vectors are equal.
DoubleVector.__eq = function (v1, v2)
  if type(v1) ~= type(v2) then
    return false
  end

  assert(type(v1) == "table" and v1["get"] and v1["len"])
  assert(type(v2) == "table" and v2["get"] and v2["len"])

  local n = cvector.double_vector_equal(v1.vec, v2.vec)

  if n ~= 0 then
    return true
  else
    return false
  end
end

-- Create a double vector from raw vector.
local function _vector_from_raw(v)
  local vector = DoubleVector:new(0)
  cvector.double_vector_free(vector.vec)
  vector.vec = v
  return vector
end

--- Vector addition.
-- @param v1 vector or number
-- @param v2 vector or number
-- @return A new vector.
DoubleVector.__add = function (v1, v2)
  local raw = nil
  if type(v1) == "number" and type(v2) == "table" then
    assert(v2["get"] and v2["len"])
    raw = cvector.double_vector_scalar_add(v2.vec, v1)
  elseif type(v1) == "table" and type(v2) == "number" then
    assert(v1["get"] and v1["len"])
    raw = cvector.double_vector_scalar_add(v1.vec, v2)
  else
    assert(type(v1) == "table" and v1["get"] and v1["len"])
    assert(type(v2) == "table" and v2["get"] and v2["len"])

    local len1 = v1:len()
    local len2 = v2:len()
    assert(len1 == len2)

    raw = cvector.double_vector_add(v1.vec, v2.vec)
  end

  return _vector_from_raw(raw)
end

--- Vector substration.
-- @param v1 vector or number
-- @param v2 vector or number
-- @return A new vector.
DoubleVector.__sub = function (v1, v2)
  local raw = nil
  if type(v1) == "number" and type(v2) == "table" then
    assert(v2["get"] and v2["len"])
    raw = cvector.double_vector_scalar_sub_first(v1, v2.vec)
  elseif type(v1) == "table" and type(v2) == "number" then
    assert(v1["get"] and v1["len"])
    raw = cvector.double_vector_scalar_sub_second(v1.vec, v2)
  else
    assert(type(v1) == "table" and v1["get"] and v1["len"])
    assert(type(v2) == "table" and v2["get"] and v2["len"])

    local len1 = v1:len()
    local len2 = v2:len()
    assert(len1 == len2)

    raw = cvector.double_vector_sub(v1.vec, v2.vec)
  end

  return _vector_from_raw(raw)
end

--- Vector multiplication.
-- @param v1 vector or number
-- @param v2 vector or number
-- @return A new vector.
DoubleVector.__mul = function (v1, v2)
  local raw = nil
  if type(v1) == "number" and type(v2) == "table" then
    assert(v2["get"] and v2["len"])
    raw = cvector.double_vector_scalar_mul(v2.vec, v1)
  elseif type(v1) == "table" and type(v2) == "number" then
    assert(v1["get"] and v1["len"])
    raw = cvector.double_vector_scalar_mul(v1.vec, v2)
  else
    assert(type(v1) == "table" and v1["get"] and v1["len"])
    assert(type(v2) == "table" and v2["get"] and v2["len"])

    local len1 = v1:len()
    local len2 = v2:len()
    assert(len1 == len2)

    raw = cvector.double_vector_mul(v1.vec, v2.vec)
  end

  return _vector_from_raw(raw)
end

--- Vector division.
-- @param v1 vector or number
-- @param v2 vector or number
-- @return A new vector.
DoubleVector.__div = function (v1, v2)
  local raw = nil
  if type(v1) == "number" and type(v2) == "table" then
    assert(v2["get"] and v2["len"])
    raw = cvector.double_vector_scalar_div_first(v1, v2.vec)
  elseif type(v1) == "table" and type(v2) == "number" then
    assert(v1["get"] and v1["len"])
    raw = cvector.double_vector_scalar_div_second(v1.vec, v2)
  else
    assert(type(v1) == "table" and v1["get"] and v1["len"])
    assert(type(v2) == "table" and v2["get"] and v2["len"])

    local len1 = v1:len()
    local len2 = v2:len()
    assert(len1 == len2)

    raw = cvector.double_vector_div(v1.vec, v2.vec)
  end

  return _vector_from_raw(raw)
end

--- Vector power.
-- @param v1 vector or number
-- @param v2 vector or number
-- @return A new vector.
DoubleVector.__pow = function (v1, v2)
  local raw = nil
  if type(v1) == "number" and type(v2) == "table" then
    assert(v2["get"] and v2["len"])
    raw = cvector.double_vector_scalar_pow_first(v1, v2.vec)
  elseif type(v1) == "table" and type(v2) == "number" then
    assert(v1["get"] and v1["len"])
    raw = cvector.double_vector_scalar_pow_second(v1.vec, v2)
  else
    assert(type(v1) == "table" and v1["get"] and v1["len"])
    assert(type(v2) == "table" and v2["get"] and v2["len"])

    local len1 = v1:len()
    local len2 = v2:len()
    assert(len1 == len2)

    raw = cvector.double_vector_pow(v1.vec, v2.vec)
  end

  return _vector_from_raw(raw)
end

--- Create a double vector.
-- @param size the size of the vector
-- @return A vector.
function DoubleVector:new(size)
  self = {}
  setmetatable(self, DoubleVector)
  self.vec = cvector.double_vector_new(size)
  return self
end

--- Create a double vector with the elements from an array-style table.
-- @param t an array-style table
-- @return A vector.
function DoubleVector:from_table(t)
  local len = #t
  local vector = DoubleVector:new(len)

  for i = 1, len do
    vector:set(i, t[i])
  end

  return vector
end

--- Index the vector.
-- @param i the index
-- @return The element.
function DoubleVector:get(index)
  return cvector.double_vector_get(self.vec, index - 1)
end

--- Assign an element by indexing the vector.
-- @param index the index
-- @param data the element
function DoubleVector:set(index, data)
  cvector.double_vector_set(self.vec, index - 1, data)
end

--- The length of the vector.
-- @return Length.
function DoubleVector:len()
  return tonumber(cvector.double_vector_size(self.vec))
end

--- The magnitude of the vector.
-- @return Magnitude.
function DoubleVector:magnitude()
  return cvector.double_vector_magnitude(self.vec)
end

--- Dot operation on two vectors
-- @param v a vector
-- @return The result
function DoubleVector:dot(v)
  assert(type(v) == "table" and v["get"] and v["len"])
  local len1 = self:len()
  local len2 = v:len()
  assert(len1 == len2)

  if type(v.vec) == "cdata" then
    return cvector.double_vector_dot(self.vec, v.vec)
  end

  local result = 0
  for i = 1, len1 do
    result = result + self:get(i) * v:get(i)
  end

  return result
end

return DoubleVector

File: src/algo/doubleVector.c
#include <stdio.h>
#include <stdlib.h>
#include <stddef.h>
#include "doubleVector.h"

char dtoa(int d) {
  if (d == 0) {
    return '0';
  } else if (d == 1) {
    return '1';
  } else if (d == 2) {
    return '2';
  } else if (d == 3) {
    return '3';
  } else if (d == 4) {
    return '4';
  } else if (d == 5) {
    return '5';
  } else if (d == 6) {
    return '6';
  } else if (d == 7) {
    return '7';
  } else if (d == 8) {
    return '8';
  } else if (d == 9) {
    return '9';
  } else {
    return 'x';
  }
}

char* utoa(size_t s) {
  size_t d = 0;
  size_t temp = s;
  while (temp > 0) {
    temp = temp / 10;
    d = d + 1;

  }

  char* str = malloc((d + 1) * sizeof(char));

  temp = s;
  int i = d;
  while (temp > 0) {
    int n = temp % 10;
    char c = dtoa(n);
    str[i - 1] = c;
    temp = temp / 10;
    i = i - 1;
  }
  str[d] = '\0';

  return str;
}

DoubleVector* double_vector_new(size_t s) {
  DoubleVector* vector = malloc(sizeof(DoubleVector));
  vector->size = s;
  vector->vec = malloc(vector->size * sizeof(double));

  for (int i = 0; i < vector->size; i++) {
    vector->vec[i] = 0.0;
  }

  return vector;
}

size_t double_vector_size(DoubleVector* v) {
  return v->size;
}

double double_vector_get(DoubleVector* v, size_t index) {
  if (index > v->size - 1) {
    double_vector_error(v, "Index out of range\n");
  }

  return v->vec[index];
}

void double_vector_set(DoubleVector* v, size_t index, double data) {
  if (index > v->size - 1) {
    double_vector_error(v, "Index out of range\n");
  }

  v->vec[index] = data;
}

int double_vector_equal(DoubleVector* v1, DoubleVector* v2) {
  if (!(v1->size == v2->size)) {
    return 0;
  }

  size_t len = v1->size;
  for (int i = 0; i < len; i++) {
    if (!(v1->vec[i] == v2->vec[i])) {
      return 0;
    }
  }

  return 1;
}

DoubleVector* double_vector_add(DoubleVector* v1, DoubleVector* v2) {
  size_t len1 = v1->size;
  size_t len2 = v2->size;
  if (len1 != len2) {
    return NULL;
  }

  DoubleVector* v = double_vector_new(len1);

  for (int i = 0; i < len1; i++) {
    double_vector_set(v, i,
      double_vector_get(v1, i) + double_vector_get(v2, i));
  }

  return v;
}

DoubleVector* double_vector_scalar_add(DoubleVector* v1, double s) {
  size_t len = v1->size;

  DoubleVector* v = double_vector_new(len);

  for (int i = 0; i < len; i++) {
    double_vector_set(v, i,
      double_vector_get(v1, i) + s);
  }

  return v;
}

DoubleVector* double_vector_sub(DoubleVector* v1, DoubleVector* v2) {
  size_t len1 = v1->size;
  size_t len2 = v2->size;
  if (len1 != len2) {
    return NULL;
  }

  DoubleVector* v = double_vector_new(len1);

  for (int i = 0; i < len1; i++) {
    double_vector_set(v, i,
      double_vector_get(v1, i) - double_vector_get(v2, i));
  }

  return v;
}

DoubleVector* double_vector_scalar_sub_first(double s, DoubleVector* v1) {
  size_t len = v1->size;

  DoubleVector* v = double_vector_new(len);

  for (int i = 0; i < len; i++) {
    double_vector_set(v, i,
      s - double_vector_get(v1, i));
  }

  return v;
}

DoubleVector* double_vector_scalar_sub_second(DoubleVector* v1, double s) {
  size_t len = v1->size;

  DoubleVector* v = double_vector_new(len);

  for (int i = 0; i < len; i++) {
    double_vector_set(v, i,
      double_vector_get(v1, i) - s);
  }

  return v;
}

DoubleVector* double_vector_mul(DoubleVector* v1, DoubleVector* v2) {
  size_t len1 = v1->size;
  size_t len2 = v2->size;
  if (len1 != len2) {
    return NULL;
  }

  DoubleVector* v = double_vector_new(len1);

  for (int i = 0; i < len1; i++) {
    double_vector_set(v, i,
      double_vector_get(v1, i) * double_vector_get(v2, i));
  }

  return v;
}

DoubleVector* double_vector_scalar_mul(DoubleVector* v1, double s) {
  size_t len = v1->size;

  DoubleVector* v = double_vector_new(len);

  for (int i = 0; i < len; i++) {
    double_vector_set(v, i,
      double_vector_get(v1, i) * s);
  }

  return v;
}

DoubleVector* double_vector_div(DoubleVector* v1, DoubleVector* v2) {
  size_t len1 = v1->size;
  size_t len2 = v2->size;
  if (len1 != len2) {
    return NULL;
  }

  DoubleVector* v = double_vector_new(len1);

  for (int i = 0; i < len1; i++) {
    double_vector_set(v, i,
      double_vector_get(v1, i) / double_vector_get(v2, i));
  }

  return v;
}

DoubleVector* double_vector_scalar_div_first(double s, DoubleVector* v1) {
  size_t len = v1->size;

  DoubleVector* v = double_vector_new(len);

  for (int i = 0; i < len; i++) {
    double_vector_set(v, i,
      s / double_vector_get(v1, i));
  }

  return v;
}

DoubleVector* double_vector_scalar_div_second(DoubleVector* v1, double s) {
  size_t len = v1->size;

  DoubleVector* v = double_vector_new(len);

  for (int i = 0; i < len; i++) {
    double_vector_set(v, i,
      double_vector_get(v1, i) / s);
  }

  return v;
}

void double_vector_error(DoubleVector* v, const char* msg) {
  fprintf(stderr, "%s", msg);
  double_vector_free(v);
  exit(EXIT_FAILURE);
}

void double_vector_free(DoubleVector* v) {
  free((void*) ((DoubleVector*) v)->vec);
  free((void*) v);
}
